### ISSUE TITLE
black

### ISSUE THREAD

**Issue by martinthomson** [ietf-tools/xml2rfc#844]
Consider applying and then enforcing the use of [black](https://pypi.org/project/black/) for this project.

**Comment by kesara** [comment]
Applying black as is changes 14837 lines.

**Comment by martinthomson** [comment]
SGTM

**Comment by kesara** [comment]
There was a similar discussion about datatracker: https://github.com/ietf-tools/datatracker/discussions/3661

Another thing to consider is main source files of xml2rfc are on the larger side, so reviewing style changes might be harder.

```wc -l xml2rfc/writers/*.py
      24 xml2rfc/writers/__init__.py
    2291 xml2rfc/writers/base.py
     354 xml2rfc/writers/doc.py
      51 xml2rfc/writers/expand.py
      39 xml2rfc/writers/expanded_xml.py
    3032 xml2rfc/writers/html.py
     864 xml2rfc/writers/legacy_html.py
     345 xml2rfc/writers/nroff.py
     433 xml2rfc/writers/paginated_txt.py
     209 xml2rfc/writers/pdf.py
    2419 xml2rfc/writers/preptool.py
    1453 xml2rfc/writers/raw_txt.py
    4968 xml2rfc/writers/text.py
     188 xml2rfc/writers/unprep.py
    1202 xml2rfc/writers/v2v3.py
   17872 total
```

**Comment by martinthomson** [comment]
Why would you review it? You can obviously run tests, but black is a very well tested tool that can basically guarantee no functional changes.  For similar projects I have seen format changes are "reviewed" by looking for bad changes, but that only requires a skim over the changes to find.

For example, we moved to using clang-format for a large project in C, clang-format is unreliable in some ways, so that needed a sanity check. Same might apply for a middling sized project like this, but I don't think that black will have similar issues.

**Comment by cabo** [comment]
Archeology in older versions (git blame and friends) plays a strong role when working on xml2rfc.
Wholesale reformatting would destroy this.
Is `black` different from wholesale reformatting?

**Comment by martinthomson** [comment]
Archaeology is a classic concern here. Given that reformatting is - or can be - a single event, it is easy to step past when on an archaeological expedition. I believe that there are even ways to automate skipping commits in blame, though Mozilla did end up disabling those in favour of a manual approach (for reasons I don't know).

Ultimately, the future is bigger than the past so this tends not to matter much in my long experience with this sort of change. Also, it makes contributing to a project easier, especially for new contributors.

**Comment by jrlevine** [comment]
I don't understand what problem bulk reformatting at this point would solve. I realize it won't introduce new bugs, but it also won't fix old bugs.
I doubly don't understand the point without having some way to enforce it on new code.

**Comment by rjsparks** [comment]
With the datatracker, we left it to use black conventions on new code but not to rewrite the codebase. I've studied some projects that have done the massive rewrite, and most have had trouble with history later - the "single commit to look past" is a gross oversimplification of what it does to toolchains.

Pushing the "what coding style to use" off to someone else is great - I don't think we have much objection to generally moving towards what black suggests. But a mass rewrite is not going to be ok.

**Comment by martinthomson** [comment]
> I don't understand what problem bulk reformatting at this point would solve.

Having just contributed a few patches to xml2rfc, I can say that there are two things that are very annoying.

1. Having to run yestests before starting.
2. Having to manually match new code style to the code around it.  For me, this includes trailing whitespace, quote choice, line wrapping, and a bunch of other minor annoyances that a code formatter usually just takes care of.

Both of these are massive time sinks and a barrier to me making further contributions (take that as a positive if you like).  Code formatting is by far the worst relative to the cost, especially when the fix is so trivial.
